routers/repo/issue.go

```diff
@@ -966,7 +966,10 @@ func ViewIssue(ctx *context.Context) {
 			ctx.Data["IsBlockedByRejection"] = pull.ProtectedBranch.MergeBlockedByRejectedReview(pull)
 			ctx.Data["GrantedApprovals"] = cnt
 		}
-		ctx.Data["IsPullBranchDeletable"] = canDelete && pull.HeadRepo != nil && git.IsBranchExist(pull.HeadRepo.RepoPath(), pull.HeadBranch)
+		ctx.Data["IsPullBranchDeletable"] = canDelete &&
+			pull.HeadRepo != nil &&
+			git.IsBranchExist(pull.HeadRepo.RepoPath(), pull.HeadBranch) &&
+			(!pull.HasMerged || ctx.Data["HeadBranchCommitID"] == ctx.Data["PullHeadCommitID"])
```

Comment:
I'm not sure I understand this condition. The branch is deletable if it _hasn't_ been merged or it's the same commit as the head?

Reply:
Yup only check that the headbranchcommitid and pullheadcommitid are the same if you've merged the PR.

it's only after merging that the two heads stop being kept in sync. Any difference in a non-merged PR could just be a race. 

I think we should probably stop updating pull heads when the PR is closed in which case we could add a third condition.

Reply:
Isn't it then?:
```
ctx.Data["IsPullBranchDeletable"] = canDelete &&
			pull.HeadRepo != nil &&
			git.IsBranchExist(pull.HeadRepo.RepoPath(), pull.HeadBranch) &&
			pull.HasMerged &&
			ctx.Data["HeadBranchCommitID"] == ctx.Data["PullHeadCommitID"]
```

Reply:
Nope because then you can't delete unless the pull has merged.

Reply:
It would have to be:

```go
ctx.Data["IsPullBranchDeletable"] = canDelete &&
			pull.HeadRepo != nil &&
			git.IsBranchExist(pull.HeadRepo.RepoPath(), pull.HeadBranch) &&
			!(pull.HasMerged &&
			ctx.Data["HeadBranchCommitID"] != ctx.Data["PullHeadCommitID"])
```

which is no easier to understand TBH.

Reply:
> 
> 
> Nope because then you can't delete unless the pull has merged.

Then I misunderstood the purpose. I thought that's exactly what we wanted. Shouldn't the PR be closed at least in order to delete its branch?

Reply:
That's dealt with at the template level:

https://github.com/go-gitea/gitea/blob/13175c012c5ed73cc600457196e57cf12c42491d/templates/repo/issue/view_content/pull.tmpl#L52-L82

and so on...

Reply:
Since a merged PR is also a closed PR, I'd add `IsClosed` as a condition, if only for consistency. Otherwise, LGTM.